### DISCUSSION TITLE
Reusing the project, modifying it and publishing it

### DISCUSSION THREAD

**Discussion by zeleny-ondrej** [ultralytics/yolov5#8911] [Q&A]
Hello,
I'm writing a paper where i use simplified and modified version of this algorithm and now i need to properly label it with the GPL3 so its all legal.

I did not really find any step by step guild how to do it, but so far i understood that the whole project has to be labeled as GPL3 licenced. So far, I've been adding links to the docstring of each function or class and noted it as either 'source:' or 'modified version of:' which i understand is one of the requirements of GPL3.

Did anybody dealt with this and could help me how to properly cite GPL3 protected sources ?

Thank you.

**Comment by glenn-jocher** [comment]
@zeleny-ondrej if you'd like to license your code as GPL 3.0 the simplest way is just to include a LICENSE file in the root of your GitHub repo, i.e. you can use the same LICENSE file here. No docstring updates are required.
https://github.com/ultralytics/yolov5/blob/master/LICENSE

At the top of the license link you'll also see some useful information:
<img width="935" alt="Screen Shot 2022-08-09 at 10 46 34 PM" src="https://user-images.githubusercontent.com/26833433/183757648-15748d80-28aa-4956-8f08-924a32f14f6d.png">

**Reply by zeleny-ondrej** [reply]
@glenn-jocher im not sure if we understood each other. Most of the code(functions, classes) in my project is either the same or modified version of the code from this yolov5 repository. I've added the license file to my project which is the first condition. 

Now i need to state the source and im not sure how. Do i need to add link for a source to each function and class or is a link at the top of the file just fine ? the functions in my projects are distributed into differente file than it is in the original project (this repo).

And at last, where to state the changes done to the project, is it okay to state it in the docstring of the code file or do the changes have to be stated in each function/class docstring. Or can it be writen in separate txt file ?

thank you for your help

**Reply by zeleny-ondrej** [reply]
update, i've searched and asked around and did following.

1.  added license file
2. added the following docstring to the top of each file

>     GPL-3.0-only License
> 
> 
>     Following code contain original or modified version of functions and
>     classes from the original implementation of YOLOv5 by Glenn Jocher
>     https://github.com/ultralytics/yolov5
> 
>     The project as a whole is a derivative work from the original
>     implementation which focuses on performance in speed and well as
>     detection performance. The code was also modified in order to be
>     compatible with Python 3.6 and run on Jetson Nano 2GB.
> 
>     Disclaimer:
>     ""
>     This program is free software: you can redistribute it and/or modify
>     it under the terms of the GNU General Public License as published by
>     the Free Software Foundation, either version 3 of the License, or
>     (at your option) any later version.
> 
>     This program is distributed in the hope that it will be useful,
>     but WITHOUT ANY WARRANTY; without even the implied warranty of
>     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
>     GNU General Public License for more details.
> 
>     You should have received a copy of the GNU General Public License
>     along with this program.  If not, see <http://www.gnu.org/licenses/>.
>     ""

do you think thats is sufficient ?

**Reply by glenn-jocher** [reply]
@zeleny-ondrej looks perfect!
